File: backend/src/models/project.py
"""
Project data models for PDLS backend
"""
from datetime import datetime
from enum import Enum
from typing import Optional, List
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
import uuid

from .base import BaseModel


class ProjectStatus(str, Enum):
    """Project status enumeration"""
    PLANNING = "planning"
    ACTIVE = "active"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    ARCHIVED = "archived"


class ProjectPriority(str, Enum):
    """Project priority enumeration"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ProjectType(str, Enum):
    """Project type enumeration"""
    SOFTWARE = "software"
    WEB = "web"
    MOBILE = "mobile"
    DATA_SCIENCE = "data_science"
    INFRASTRUCTURE = "infrastructure"
    RESEARCH = "research"
    OTHER = "other"


class ProjectMemberRole(str, Enum):
    """Project member role enumeration"""
    OWNER = "owner"
    ADMIN = "admin"
    MEMBER = "member"
    VIEWER = "viewer"


class InvitationStatus(str, Enum):
    """Project invitation status enumeration"""
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    CANCELLED = "cancelled"
    EXPIRED = "expired"


class Project(BaseModel):
    """Project model representing development projects"""
    __tablename__ = "projects"

    id = Column(Integer, primary_key=True, index=True)
    uuid = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, index=True)
    
    # Basic Information
    name = Column(String(200), nullable=False, index=True)
    description = Column(Text, nullable=True)
    short_description = Column(String(500), nullable=True)
    
    # Project Classification
    project_type = Column(SQLEnum(ProjectType), default=ProjectType.SOFTWARE, nullable=False)
    status = Column(SQLEnum(ProjectStatus), default=ProjectStatus.PLANNING, nullable=False)
    priority = Column(SQLEnum(ProjectPriority), default=ProjectPriority.MEDIUM, nullable=False)
    
    # Project Metadata
    project_code = Column(String(50), unique=True, nullable=True, index=True)  # e.g., "PDLS-2024"
    version = Column(String(50), default="1.0.0", nullable=False)
    tags = Column(String(500), nullable=True)  # JSON string of tags
    
    # Dates and Timeline
    start_date = Column(DateTime, nullable=True)
    end_date = Column(DateTime, nullable=True)
    estimated_hours = Column(Integer, nullable=True)
    actual_hours = Column(Integer, default=0)
    
    # Repository and External Links
    repository_url = Column(String(500), nullable=True)
    documentation_url = Column(String(500), nullable=True)
    deployment_url = Column(String(500), nullable=True)
    
    # Project Settings
    is_public = Column(Boolean, default=False, nullable=False)
    allow_external_contributors = Column(Boolean, default=False, nullable=False)
    require_approval_for_logs = Column(Boolean, default=False, nullable=False)
    
    # Owner Information
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    owner = relationship("User", foreign_keys=[owner_id], back_populates="owned_projects")
    creator = relationship("User", foreign_keys=[created_by], back_populates="created_projects")
    members = relationship("ProjectMember", back_populates="project", cascade="all, delete-orphan")
    # development_logs = relationship("DevelopmentLog", back_populates="project", cascade="all, delete-orphan")  # TODO: Add when DevelopmentLog model is created
    project_files = relationship("ProjectFile", back_populates="project", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Project(id={self.id}, name='{self.name}', status='{self.status}')>"


class ProjectMember(BaseModel):
    """Project member association model"""
    __tablename__ = "project_members"

    id = Column(Integer, primary_key=True, index=True)
    
    # Foreign Keys
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Member Information
    role = Column(SQLEnum(ProjectMemberRole), default=ProjectMemberRole.MEMBER, nullable=False)
    is_active = Column(Boolean, default=True, nullable=False)
    
    # Permissions
    can_create_logs = Column(Boolean, default=True, nullable=False)
    can_edit_logs = Column(Boolean, default=False, nullable=False)
    can_delete_logs = Column(Boolean, default=False, nullable=False)
    can_manage_members = Column(Boolean, default=False, nullable=False)
    can_edit_project = Column(Boolean, default=False, nullable=False)
    
    # Timestamps
    joined_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    project = relationship("Project", back_populates="members")
    user = relationship("User", back_populates="project_memberships")

    def __repr__(self):
        return f"<ProjectMember(project_id={self.project_id}, user_id={self.user_id}, role='{self.role}')>"


class ProjectFile(BaseModel):
    """Project file model for document and asset management"""
    __tablename__ = "project_files"

    id = Column(Integer, primary_key=True, index=True)
    
    # Foreign Keys
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    uploaded_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # File Information
    filename = Column(String(255), nullable=False)
    original_filename = Column(String(255), nullable=False)
    file_path = Column(String(500), nullable=False)  # MinIO path
    file_size = Column(Integer, nullable=False)  # Size in bytes
    mime_type = Column(String(100), nullable=False)
    
    # File Metadata
    description = Column(Text, nullable=True)
    category = Column(String(50), nullable=True)  # e.g., "documentation", "assets", "code"
    tags = Column(String(500), nullable=True)  # JSON string of tags
    
    # File Status
    is_public = Column(Boolean, default=False, nullable=False)
    is_archived = Column(Boolean, default=False, nullable=False)
    
    # Timestamps
    uploaded_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    project = relationship("Project", back_populates="project_files")
    uploader = relationship("User", back_populates="uploaded_project_files")

    def __repr__(self):
        return f"<ProjectFile(id={self.id}, filename='{self.filename}', project_id={self.project_id})>"


class ProjectInvitation(BaseModel):
    """Project invitation model for member invitations"""
    __tablename__ = "project_invitations"

    id = Column(Integer, primary_key=True, index=True)
    uuid = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, index=True)
    
    # Foreign Keys
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    inviter_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    invited_user_id = Column(Integer, ForeignKey("users.id"), nullable=True)  # Can be null for email invitations
    
    # Invitation Information
    email = Column(String(255), nullable=False, index=True)
    role = Column(SQLEnum(ProjectMemberRole), default=ProjectMemberRole.MEMBER, nullable=False)
    message = Column(Text, nullable=True)
    
    # Invitation Status
    status = Column(String(20), default="pending", nullable=False)  # pending, accepted, rejected, expired
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    expires_at = Column(DateTime, nullable=False)  # 7 days from creation
    responded_at = Column(DateTime, nullable=True)
    
    # Relationships
    project = relationship("Project")
    inviter = relationship("User", foreign_keys=[inviter_id])
    invited_user = relationship("User", foreign_keys=[invited_user_id])

    def __repr__(self):
        return f"<ProjectInvitation(id={self.id}, email='{self.email}', status='{self.status}')>"

File: frontend/src/types/project.ts
/**
 * Project Management Types for Phase 5
 * Detailed type definitions for project management components
 */

import { BaseEntity, ID, UUID, Timestamp, Email, URL } from './index';

// ============= Project Enums from Backend =============

export enum ProjectType {
  SOFTWARE = 'software',
  WEB = 'web',
  MOBILE = 'mobile',
  DATA_SCIENCE = 'data_science',
  INFRASTRUCTURE = 'infrastructure',
  RESEARCH = 'research',
  OTHER = 'other'
}

export enum ProjectStatus {
  PLANNING = 'planning',
  ACTIVE = 'active',
  ON_HOLD = 'on_hold',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  ARCHIVED = 'archived'
}

export enum ProjectPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical'
}

export enum ProjectMemberRole {
  OWNER = 'owner',
  ADMIN = 'admin',
  MEMBER = 'member',
  VIEWER = 'viewer'
}

export enum InvitationStatus {
  PENDING = 'pending',
  ACCEPTED = 'accepted',
  DECLINED = 'declined',
  CANCELLED = 'cancelled',
  EXPIRED = 'expired'
}

// ============= Core Project Interfaces =============

export interface Project extends BaseEntity {
  uuid: UUID;
  name: string;
  description?: string;
  project_type: ProjectType;
  status: ProjectStatus;
  priority: ProjectPriority;
  start_date?: Timestamp;
  end_date?: Timestamp;
  repository_url?: URL;
  documentation_url?: URL;
  tags: string[];
  
  // Ownership
  owner_id: ID;
  created_by: ID;
  
  // Statistics
  member_count: number;
  file_count: number;
  
  // Relationships (populated when needed)
  owner?: ProjectUser;
  creator?: ProjectUser;
  members?: ProjectMember[];
  files?: ProjectFile[];
  recent_activity?: ProjectActivity[];
}

export interface ProjectUser {
  id: ID;
  username: string;
  email: Email;
  full_name?: string;
  avatar_url?: URL;
  role?: string; // User role in system
}

export interface ProjectMember extends BaseEntity {
  project_id: ID;
  user_id: ID;
  role: ProjectMemberRole;
  joined_at: Timestamp;
  
  // Relationships
  user?: ProjectUser;
  project?: Project;
}

export interface ProjectFile extends BaseEntity {
  project_id: ID;
  filename: string;
  original_filename: string;
  file_size: number;
  content_type: string;
  file_path: string;
  upload_url?: URL;
  download_url?: URL;
  
  // Ownership
  uploaded_by: ID;
  uploader?: ProjectUser;
}

export interface ProjectInvitation extends BaseEntity {
  project_id: ID;
  inviter_id: ID;
  invited_user_id?: ID;
  invited_email?: Email;
  role: ProjectMemberRole;
  status: InvitationStatus;
  message?: string;
  expires_at: Timestamp;
  
  // Relationships
  project?: Project;
  inviter?: ProjectUser;
  invited_user?: ProjectUser;
}

export interface ProjectActivity {
  id: ID;
  project_id: ID;
  user_id: ID;
  activity_type: 'project_created' | 'project_updated' | 'member_added' | 'member_removed' | 'file_uploaded' | 'invitation_sent';
  title: string;
  description: string;
  timestamp: Timestamp;
  
  // Relationships
  user?: ProjectUser;
  metadata?: Record<string, any>;
}

// ============= Create/Update DTOs =============

export interface CreateProjectRequest {
  name: string;
  description?: string;
  project_type: ProjectType;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  start_date?: Timestamp;
  end_date?: Timestamp;
  repository_url?: URL;
  documentation_url?: URL;
  tags?: string[];
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  project_type?: ProjectType;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  start_date?: Timestamp;
  end_date?: Timestamp;
  repository_url?: URL;
  documentation_url?: URL;
  tags?: string[];
}

export interface CreateInvitationRequest {
  project_id: ID;
  invited_user_id?: ID;
  invited_email?: Email;
  role: ProjectMemberRole;
  message?: string;
}

export interface UpdateMemberRoleRequest {
  role: ProjectMemberRole;
}

export interface FileUploadRequest {
  file: File;
  project_id: ID;
  description?: string;
}

// ============= Query and Filter Types =============

export interface ProjectListQuery {
  page?: number;
  page_size?: number;
  search?: string;
  project_type?: ProjectType;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  owner_id?: ID;
  member_id?: ID;
  tags?: string[];
  sort_by?: 'name' | 'created_at' | 'updated_at' | 'status' | 'priority';
  sort_order?: 'asc' | 'desc';
}

export interface ProjectMemberQuery {
  project_id: ID;
  role?: ProjectMemberRole;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface ProjectInvitationQuery {
  project_id?: ID;
  inviter_id?: ID;
  invited_user_id?: ID;
  status?: InvitationStatus;
  page?: number;
  page_size?: number;
}

export interface ProjectFileQuery {
  project_id: ID;
  search?: string;
  content_type?: string;
  uploaded_by?: ID;
  page?: number;
  page_size?: number;
  sort_by?: 'filename' | 'created_at' | 'file_size';
  sort_order?: 'asc' | 'desc';
}

// ============= Component State Types =============

export interface ProjectListState {
  projects: Project[];
  loading: boolean;
  error?: string;
  query: ProjectListQuery;
  total: number;
  selectedProjects: Project[];
}

export interface ProjectDetailState {
  project?: Project;
  members: ProjectMember[];
  files: ProjectFile[];
  invitations: ProjectInvitation[];
  activity: ProjectActivity[];
  loading: {
    project: boolean;
    members: boolean;
    files: boolean;
    invitations: boolean;
    activity: boolean;
  };
  error?: string;
}

export interface ProjectFormState {
  data: CreateProjectRequest;
  errors: Partial<Record<keyof CreateProjectRequest, string>>;
  loading: boolean;
  mode: 'create' | 'edit';
}

export interface InvitationFormState {
  data: CreateInvitationRequest;
  errors: Partial<Record<keyof CreateInvitationRequest, string>>;
  loading: boolean;
  searchResults: ProjectUser[];
  searchLoading: boolean;
}

export interface FileUploadState {
  files: Array<{
    file: File;
    progress: number;
    status: 'pending' | 'uploading' | 'completed' | 'error';
    error?: string;
    result?: ProjectFile;
  }>;
  uploading: boolean;
}

// ============= API Response Types =============

export interface ProjectListResponse {
  projects: Project[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ProjectDetailResponse extends Project {
  members: ProjectMember[];
  recent_files: ProjectFile[];
  recent_activity: ProjectActivity[];
  statistics: {
    total_members: number;
    total_files: number;
    files_size: number;
    last_activity: Timestamp;
  };
}

export interface MemberListResponse {
  members: ProjectMember[];
  total: number;
  page: number;
  page_size: number;
}

export interface InvitationListResponse {
  invitations: ProjectInvitation[];
  total: number;
  page: number;
  page_size: number;
}

export interface FileListResponse {
  files: ProjectFile[];
  total: number;
  total_size: number;
  page: number;
  page_size: number;
}

export interface ActivityListResponse {
  activities: ProjectActivity[];
  total: number;
  page: number;
  page_size: number;
}

// ============= UI Component Props Types =============

export interface ProjectCardProps {
  project: Project;
  onClick?: (project: Project) => void;
  onEdit?: (project: Project) => void;
  onDelete?: (project: Project) => void;
  showActions?: boolean;
  compact?: boolean;
}

export interface ProjectListProps {
  projects: Project[];
  loading?: boolean;
  onProjectSelect?: (project: Project) => void;
  onProjectEdit?: (project: Project) => void;
  onProjectDelete?: (project: Project) => void;
  onRefresh?: () => void;
  query?: ProjectListQuery;
  onQueryChange?: (query: ProjectListQuery) => void;
}

export interface ProjectFormProps {
  project?: Project;
  onSubmit: (data: CreateProjectRequest | UpdateProjectRequest) => void;
  onCancel?: () => void;
  loading?: boolean;
  errors?: Partial<Record<keyof CreateProjectRequest, string>>;
}

export interface ProjectDetailProps {
  projectId: ID;
  project?: Project;
  onProjectUpdate?: (project: Project) => void;
}

export interface MemberListProps {
  projectId: ID;
  members: ProjectMember[];
  currentUserRole?: ProjectMemberRole;
  onInviteMember?: () => void;
  onRemoveMember?: (member: ProjectMember) => void;
  onUpdateRole?: (member: ProjectMember, role: ProjectMemberRole) => void;
  loading?: boolean;
}

export interface InvitationFormProps {
  projectId: ID;
  onSubmit: (data: CreateInvitationRequest) => void;
  onCancel?: () => void;
  loading?: boolean;
  errors?: Partial<Record<keyof CreateInvitationRequest, string>>;
}

export interface FileManagerProps {
  projectId: ID;
  files: ProjectFile[];
  onUpload?: (files: File[]) => void;
  onDownload?: (file: ProjectFile) => void;
  onDelete?: (file: ProjectFile) => void;
  loading?: boolean;
  uploadProgress?: FileUploadState;
}

// ============= Error Types =============

export interface ProjectError {
  code: string;
  message: string;
  field?: keyof CreateProjectRequest | keyof UpdateProjectRequest;
  details?: any;
}

export interface ValidationError {
  field: string;
  message: string;
  code?: string;
}

// ============= Permission Types =============

export interface ProjectPermissions {
  can_view: boolean;
  can_edit: boolean;
  can_delete: boolean;
  can_manage_members: boolean;
  can_invite_members: boolean;
  can_upload_files: boolean;
  can_delete_files: boolean;
  can_change_settings: boolean;
}

export interface UserProjectRole {
  project_id: ID;
  user_id: ID;
  role: ProjectMemberRole;
  permissions: ProjectPermissions;
  joined_at: Timestamp;
}

// ============= Dashboard Types =============

export interface ProjectDashboardData {
  my_projects: Project[];
  recent_projects: Project[];
  project_statistics: {
    total_projects: number;
    active_projects: number;
    projects_as_owner: number;
    projects_as_member: number;
  };
  recent_activity: ProjectActivity[];
  upcoming_deadlines: Array<{
    project: Project;
    end_date: Timestamp;
    days_remaining: number;
  }>;
}

// ============= Export all project types =============
export type {
  // Core interfaces are already exported via individual exports above
};